Linting

asherber opened the pull request:
This adds linting to the build process, using luacheck. 

1. Luacheck is added to the list of recommended extensions for VSCode, so that devs will see locally the same warnings and errors that will be found on GH.
2. The GH action will run on any push, and on PRs to master; any failures will show up in the PR.
3. The `.luacheckrc` in the repo (which is used locally as well as on GH) is configured to ignore formatting issues, as well as a few other things that Robert and I felt were not important.
4. Only files in the `src` directory are linted, and third-party files in there (like `lunajson` and `mobdebug`) are ignored.
5. A dev may decide that a certain warning is acceptable and may be bypassed. For example:
    ```lua
    local region = finenv.Region()

    function do_something(region)
    ...
    end
    ```
    This will generate a warning for the function definition that the `region` parameter shadows the earlier variable with the same name. To suppress the warning, add a comment `-- luacheck: ignore region` on the same line as the function definition. 

The docs for luacheck are [here](https://luacheck.readthedocs.io/en/stable/index.html). In particular, here are the [potential warnings](https://luacheck.readthedocs.io/en/stable/warnings.html) that can be emitted (note that our `.luacheckrc` disables some of these) and instructions for [using comments](https://luacheck.readthedocs.io/en/stable/inline.html) to disable specific warnings. Devs are expected to disable warnings as locally as possible in their code.

There were about 70 warnings already in the codebase. I went through and fixed the ones that were easy to fix, and I added ignore hints to the others. **Please take a look at your scripts** and let me know if any changes are needed.

Suggestions and comments are welcome.

asherber commented:
@cv-on-hub I don't think this actually affects any of your current scripts, but we wanted to get your eyes on the PR as well.

cv-on-hub commented:
> we wanted to get your eyes on the PR as well

Thanks @asherber. The mechanics of how this all fits together is beyond my understanding but it looks like enhanced strength for the repo.

asherber commented:
As I'm looking at some of the potential warnings, I think I'm going to re-enable some of the warnings around unused variables -- these just create clutter and make code harder to read. This is going to make some extra work for me to clean up current instances, so I'm going to set this PR back to draft and will advise when I've pushed the new commit. Feel free to continue the discussion.

asherber commented:
I take it back. This is a good first step, and we should get it merged. Further cleanup can come later.

jwink75 commented:
For anybody interested: Zerobrane Studio does indeed support Luacheck, it just requires setting a parameter in the configuration file.



> On Jan 23, 2024, at 11:59 AM, Aaron Sherber ***@***.***> wrote:
> 
> This adds linting to the build process, using luacheck.
> 
> Luacheck is added to the list of recommended extensions for VSCode, so that devs will see locally the same warnings and errors that will be found on GH.
> The GH action will run on any push, and on PRs to master; any failures will show up in the PR.
> The .luacheckrc in the repo (which is used locally as well as on GH) is configured to ignore formatting issues, as well as a few other things that Robert and I felt were not important.
> Only files in the src directory are linted, and third-party files in there (like lunajson and mobdebug) are ignored.
> A dev may decide that a certain warning is acceptable and may be bypassed. For example:
> local region = finenv.Region()
> 
> function do_something(region)
> ...
> end
> This will generate a warning for the function definition that the region parameter shadows the earlier variable with the same name. To suppress the warning, add a comment -- luacheck: ignore region on the same line as the function definition.
> The docs for luacheck are here <https://luacheck.readthedocs.io/en/stable/index.html>. In particular, here are the potential warnings <https://luacheck.readthedocs.io/en/stable/warnings.html> that can be emitted (note that our .luacheckrc disables some of these) and instructions for using comments <https://luacheck.readthedocs.io/en/stable/inline.html> to disable specific warnings. Devs are expected to disable warnings as locally as possible in their code.
> 
> There were about 70 warnings already in the codebase. I went through and fixed the ones that were easy to fix, and I added ignore hints to the others. Please take a look at your scripts and let me know if any changes are needed.
> 
> Suggestions and comments are welcome.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/finale-lua/lua-scripts/pull/655
> 
> Commit Summary
> 
> 7bdf531 <https://github.com/finale-lua/lua-scripts/pull/655/commits/7bdf531d8eec8cc40d5ee94bba0395871998063f> Add linting workflow and config
> 304616d <https://github.com/finale-lua/lua-scripts/pull/655/commits/304616d987961bae321f80a7aa0c1e3de03572a8> Luacheck hinting and fixes
> File Changes (34 files <https://github.com/finale-lua/lua-scripts/pull/655/files>)
> A .github/workflows/lint.yml <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-107e910e9f2ebfb9a741fa10b2aa7100cc1fc4f5f3aca2dfe78b905cbd73c0d2> (16)
> A .luacheckrc <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-69ff5cbc209d68db600cc536c6b9ec13b057b55207825d17e71f0ae0d6b0b3ff> (37)
> M .vscode/extensions.json <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-c16655a98a3ee89a7636a59c59a72b0e93649e3a1e947327cfc43a1336b4e912> (2)
> M .vscode/recommended_settings.json <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-ed5a0a54a184cda00eb28983714932b47455b28d806b44097c8efd4476782d17> (1)
> M src/UI_switch_to_selected_part.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-08401d37b7f15b5f03fb5266f71b4efca3c527696c08a6385525da99284fdd62> (8)
> M src/chord_accidental_adjust_down.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-feb2f261c01ed4e569ac5fc5e2acf432ebabb53aa23bf92898aeac9b3f69d147> (8)
> M src/chord_accidental_adjust_up.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-81b6a8e390240073877d9594748e4ac46300eb9afea294aceed962c4b2a30ed3> (8)
> M src/document_options_as_json.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-eb29604186463730e3a99cd032f0bcde10e25e3434f74b6e2faffdd6a6009b3c> (5)
> M src/dynamics_move_above_staff.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-8b01ba34dcb3b332d4ca59b04d11e329d7c7ae52d8acaa74d7c21cc6d5683f88> (6)
> M src/finale_lua_menu_organizer.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-7e815ac6c92129a869791a113dbcc60f93730abadf19924539144c3e198887a0> (3)
> M src/harp_pedal_wizard.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-c64acefda21bb3939d10e4f3bcddb8603274e793101a53cfe5027a0adce60d4d> (31)
> M src/layers_create_whole_rests_1_2.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-b3b5e47eff921af4d865643b0037a8016b8751cf2dc08db11d86a1964d4c541b> (4)
> M src/library/articulation.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-34c64829e26dfe9e9ffe99e3c0f5d291eb01d354489872e89c592ab8eb68bf14> (1)
> M src/library/clef.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-efa3a9de7c091a3b2b0c26d3b2688917462e8c1021009cf6535a19cf0f62e22f> (5)
> M src/library/general_library.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-ca92f2967fe48a516123919b227d2245d8fd931773de6e8d14e60ce5d63b75b5> (4)
> M src/library/lua_compatibility.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-254ce3d90038a4d4f9c20426c12a58f1d2861cd20a2da9ca1a753c2a28afe694> (4)
> M src/library/score.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-c0386bfc3d4d4239cdae79d3dd3a0bb1610a6c4f983c4d1d2358efc9d3c8f049> (4)
> M src/library/smartshape.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-050ce9fb94bde57dd3e9c2ebcf46b685836c19e2508c4d85ede18b904e36524f> (4)
> M src/lyrics_baseline_spacing.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-13b5c977a79580de5e1f48ca86270961923c19ad3605ababba9353f6aef774dc> (4)
> M src/lyrics_openai_hyphenation.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-15519e37152fed67c2aeebd830e7db32ef5b53e4a2e9dd2dcb86be5135985149> (4)
> M src/mixin/FCMControl.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-a84205d005eb5ab525350877458ba855b2593be0d418eab5456d0ba9ab8b8f7e> (2)
> M src/mixin/FCXCtrlUpDown.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-bef941de93bc9078ef232e085137ab3bca5caf2845eb0392b852c2621cd7ae20> (2)
> M src/mmrest_widen_to_tempo_mark.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-c8cef92812fc781054c7f91cd1625b6d4c4d0df0cf0dde3e5a15c72471b7970d> (4)
> M src/note_cluster_determinate.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-71c1a3d1214c4657cd5b4d7fb761344f2156fb88cb8da4bd6f22267d7cd529c9> (74)
> M src/page_format_wizard.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-d4c9f03caa8f65a1c07ebca3bb4796984b26fb8edb311dc2855c4f329209f639> (55)
> M src/playback_mute_notes_in_region.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-961ec07b12c3f049611dcf33dcb7fe06262e9c41e6c37134183c9aaa4d4579fb> (4)
> M src/playback_selected_staves.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-55059375e707f5b5727887deb7506899c014a6a69144929db03047b33c93c903> (6)
> M src/playback_unmute_all_staves.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-fe86e3ab91db89364942353ebecfb40231d8ab158935a958f3cb6202e79dd367> (6)
> M src/playback_unmute_notes_in_region.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-bb895608523761b87360a9cbbb23b37c2ced1dcf7cbc4ea1bfef19b0f18dfa7c> (4)
> M src/smartshape_move_repeat_bracket_for_chords.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-8f56900379d41bdd10e9a113fd6bbdf28a06bd5d6d7c672bde0522a10ef49c49> (4)
> M src/staff_rename.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-969ab6aef7655a874366bfac419f8375be83937663faa7d11d4efb35afc226ca> (6)
> M src/staff_split_layers.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-d9823f836ead95335d80510fe78e31b2a5b8647565b32f684a3448d32488a69b> (6)
> M src/standalone_hairpin_adjustment.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-518255e59828eaa080ad57edce7fac6dc4431b53833aca643097649713e312df> (4)
> M src/system_move.lua <https://github.com/finale-lua/lua-scripts/pull/655/files#diff-227524774f53fd79b512d69874259aee9a5ff845aae2f65e1f89df622e7139ed> (4)
> Patch Links:
> 
> https://github.com/finale-lua/lua-scripts/pull/655.patch
> https://github.com/finale-lua/lua-scripts/pull/655.diff
> —
> Reply to this email directly, view it on GitHub <https://github.com/finale-lua/lua-scripts/pull/655>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMCBUQ34UMA2AXZ4LNDJEKLYQAJCLAVCNFSM6AAAAABCHSZ7XOVHI2DSMVQWIX3LMV43ASLTON2WKOZSGA4TMOBYGEZTGNY>.
> You are receiving this because you are subscribed to this thread.
>

jwink75 commented:
I don’t know if anyone else is using Zerobrane, but I did figure out how to configure it to use Luacheck and use the luacheckrc parameters from the ‘linting’ pull request

> On Jan 23, 2024, at 1:47 PM, Aaron Sherber ***@***.***> wrote:
> 
> 
> As I'm looking at some of the potential warnings, I think I'm going to re-enable some of the warnings around unused variables -- these just create clutter and make code harder to read. This is going to make some extra work for me to clean up current instances, so I'm going to set this PR back to draft and will advise when I've pushed the new commit. Feel free to continue the discussion.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/finale-lua/lua-scripts/pull/655#issuecomment-1906973756>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMCBUQ7N5MKUJ7JU42ACUCDYQAVXDAVCNFSM6AAAAABCHSZ7XOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMBWHE3TGNZVGY>.
> You are receiving this because your review was requested.
>

asherber commented:
Okay, I have implemented stricter linting, primarily looking for unused variables. In addition, new global variables may only be initialized in the top-level scope. I have adjusted all scripts so that the linting passes.

Please review all of your scripts one more time to make sure that nothing is broken. I'd like to merge this in the next few days.

cc: @cv-on-hub @jwink75 @Nick-Mazuk @rpatters1 @CJGarciaMusic

cv-on-hub commented:
> new global variables may only be initialized in the top-level scope.

I have only one script using global vars -  `ostinato_maker.lua`. I've revised it leaving just one global, `global_dialog`, which I think is needed for modeless operation. Moved its initialisation to top-level, mimicking Robert's sample script `modeless_dialog.lua`. Submitted in #661 - deploy check successful.

jwink75 commented:
The new rules don’t like me. 

> On Jan 25, 2024, at 4:19 PM, Carl Vine ***@***.***> wrote:
> 
> 
> new global variables may only be initialized in the top-level scope.
> 
> I have only one script using global vars - ostinato_maker.lua. I've revised it leaving just one global, global_dialog, which I think is needed for modeless operation. Moved its initialisation to top-level, mimicking Robert's sample script modeless_dialog.lua. Submitted in #661 <https://github.com/finale-lua/lua-scripts/pull/661> - deploy check successful.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/finale-lua/lua-scripts/pull/655#issuecomment-1911208783>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMCBUQ5ZBZNYBQ6V75BXPDLYQLY7JAVCNFSM6AAAAABCHSZ7XOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJRGIYDQNZYGM>.
> You are receiving this because you were mentioned.
>

jwink75 commented:
So here’s a question…

There are a lot of times where I’ll do something along the lines of:

for key, val in pairs(table) do
	val = variable
end

 Because that’s a quick and easy way of going through a table, rather than figuring out how many entries there are and setting up some index variable that will increment. The linting rules really don’t like the fact that in the above code ‘key’ is unused, so I get a ton of errors.

Is there another way of iterating through a table that the linter likes but retains the simplicity of pairs or ipairs?

> On Jan 25, 2024, at 4:19 PM, Carl Vine ***@***.***> wrote:
> 
> 
> new global variables may only be initialized in the top-level scope.
> 
> I have only one script using global vars - ostinato_maker.lua. I've revised it leaving just one global, global_dialog, which I think is needed for modeless operation. Moved its initialisation to top-level, mimicking Robert's sample script modeless_dialog.lua. Submitted in #661 <https://github.com/finale-lua/lua-scripts/pull/661> - deploy check successful.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/finale-lua/lua-scripts/pull/655#issuecomment-1911208783>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMCBUQ5ZBZNYBQ6V75BXPDLYQLY7JAVCNFSM6AAAAABCHSZ7XOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJRGIYDQNZYGM>.
> You are receiving this because you were mentioned.
>

jwink75 commented:
It’s even worse than I thought, because I just got an error on:

for i = 1, 5, 1 do

Saying that I was an unused variable… But as my limited understanding of this kind of thing goes, it is NOT unused, it’s just all sort of internal to the loop (initial value is 1, go until it’s 5, increment by 1 each loop).




> On Jan 25, 2024, at 4:19 PM, Carl Vine ***@***.***> wrote:
> 
> 
> new global variables may only be initialized in the top-level scope.
> 
> I have only one script using global vars - ostinato_maker.lua. I've revised it leaving just one global, global_dialog, which I think is needed for modeless operation. Moved its initialisation to top-level, mimicking Robert's sample script modeless_dialog.lua. Submitted in #661 <https://github.com/finale-lua/lua-scripts/pull/661> - deploy check successful.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/finale-lua/lua-scripts/pull/655#issuecomment-1911208783>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AMCBUQ5ZBZNYBQ6V75BXPDLYQLY7JAVCNFSM6AAAAABCHSZ7XOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJRGIYDQNZYGM>.
> You are receiving this because you were mentioned.
>

asherber commented:
All, sorry if I wasn't clear about this -- I'm not asking you to go through and fix issues, because I've already taken care of that. Please take a look at the copies of your scripts in the [head branch for this PR] and make sure that my altered versions work correctly.

@jwink75 I know that there are a bunch of things in your scripts that the linter doesn't like. Instead of overwhelming the comments here, would you mind either contacting me externally (my email address is in my scripts) or opening a [Discussion](https://github.com/finale-lua/lua-scripts/discussions/landing) if you'd rather keep this discussion on GH?

jwink75 commented:
Happy to move to a discussion, but the link didn't work and I can't seem to find it in the GitHub UI, so forgive me one more comment here...

Aaron, I was testing my harp pedals script and there were a few things that are not working, so I pulled up an older version that is working and I'm going through and trying to address the linting issues myself... Partly as a learning exercise so I don't make those goofs going forward. Like I said above, though, there seem to be a number of things that shouldn't be errors, so I'm wondering how I can handle those better (aside from cheating the system by using luacheck: ignore)

cv-on-hub commented:
> take a look at the copies of your scripts in the `head branch`

Sorry - I also misunderstood. Your `luacheck` adjustment to that script is fine. I'll delete #661  and wait until the dust has settled for other changes.

asherber commented:
The discussions tab is right at the top of the GH UI:

![image](https://github.com/finale-lua/lua-scripts/assets/5248041/4760464d-c921-4ce0-8022-ac1ffa0f35d6)

@jwink75 The harp pedals script has lots of issues; you'll see that I addressed some of the low-hanging fruit but then marked the whole script with an ignore hint when I added the stricter rules. I'm happy do do a pair programming session with you and walk through some of what's going on. Please email me and we can set up a time that works for you.

ThistleSifter commented:
While I appreciate the convention of using an underscore for unused variables in a `for` iterator, I don't agree with enforcing it in this repo. Aside from being a convention and not a rule, it is a form of programmer's shorthand and shorthand only makes sense if you already know what it means. For those who aren't programmers, but have simply picked up enough knowledge to start writing scripts, it is more intuitive to write them out, easier to understand when returning to it in the future and harmless. We have a variety of contributors with varying backgrounds and experience in programming and I think we need to acknowledge that and work with it.

I'd like the calls to `local mixin = require("library.mixin")` at the top of each mixin file to be left in. It simplifies editing and costs nothing to have them in since the `mixin` library is what calls those files in the first place.

Please tag me in future.

ThistleSifter commented on src/library/mixin_helper.lua:
```diff
@@ -528,7 +528,7 @@ function mixin_helper.create_custom_window_change_event(...)
     local function trigger_func(window, immediate)
         if type(window) == "boolean" and window then
-            for win in event.target_iterator() do
+            for _ in event.target_iterator() do
```
This line was correct as it was. The typo is in lines 533 and 535 below which should reference `win` instead of `window`.

asherber replied:
Thanks, will fix.

asherber commented:
@ThistleSifter Thanks for your comments, and sorry I didn't tag you earlier.

I agree about supporting contributors of different levels of experience, but I also think it's worthwhile to nudge them towards better code. What I like about the underscore is that it should be the outcome of a virtuous cycle: write a loop, get a lint warning, realize that the variable isn't used, and acknowledge it as such. The lint warning itself is useful because it tells you that you're not using the variable, and maybe you meant to, in which case you can fix your loop body.

Similarly, I'm not crazy about putting those unused `local mixin` declarations back in if they're not actually being used -- but if you feel strongly about it I will (along with an ignore hint to make the linter happy).

jwink75 commented:
> The discussions tab is right at the top of the GH UI:
> 
> ![image](https://private-user-images.githubusercontent.com/5248041/299849430-4760464d-c921-4ce0-8022-ac1ffa0f35d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYyMzU2MjIsIm5iZiI6MTcwNjIzNTMyMiwicGF0aCI6Ii81MjQ4MDQxLzI5OTg0OTQzMC00NzYwNDY0ZC1jOTIxLTRjZTAtODAyMi1hYzFmZmEwZjM1ZDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjZUMDIxNTIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTczYjM5NDA5ZmI1ZTdjMDUzZTU1MGI0YmFjMTRkNGE0NWNmMTBhMWZlODc1ODkxYzkwNDdlYzgxMDM0ODQ4MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.VYaUnSG2FvAJIz5RXDRv1hVK4XDLozwVKIMsqu5LfwU)
> 

i checked again in case I was blind or crazy... The discussions button truly is NOT in my UI!
<img width="986" alt="image" src="https://github.com/finale-lua/lua-scripts/assets/50600515/cacde6e1-017b-4097-b924-ec0f1036c688">

cv-on-hub commented:
Have now checked all of my updated scripts in this PR and they're all fine except for:

- `hairpin_creator.lua`: failing to initialise `y_current` to zero at line 551 results in an error in FCMCustomWindow, as it won’t accept a nil coordinate.
- `ostinato_maker.lua`: does `global_dialog` need to be global, or is that just if Modeless is in conjunction with RetainLuaState?

asherber commented:
@cv-on-hub 

1. Can you see where the error is thrown? It looks to me like `y_current` is initialized in other places before it is used.
2. Looking at the version of your script in this branch, you reference `global_dialog` inside a function [here](https://github.com/asherber/lua-scripts/blob/70d75ad13e989ec56f9d0cdf6b7e2e432dc9208c/src/ostinato_maker.lua#L319), which implies that the variable needs to be global.

cv-on-hub commented:
> 1. Can you see where the error is thrown?

1. Yes - you're right. I was testing a slightly different version of the file that needed initialization. This one doesn't.
2. The global_dialog call at [line 319](https://github.com/asherber/lua-scripts/blob/70d75ad13e989ec56f9d0cdf6b7e2e432dc9208c/src/ostinato_maker.lua#L319) presumed that the var was global modelled on all of Robert's old Modeless scripts. But it would also work with a local top-level var, and I see that you've changed that var to top-level local on his scripts as well.

asherber commented:
@cv-on-hub I did initially change those vars to top-level local, but after discussion with Robert I changed it back to global and added the ignore hint. Either way is fine from a linter point of view. If the var is local then it won't persist across invocations, but it sounds like you may not want that behavior anyway.

rpatters1 commented:
@cv-on-hub The `global_dialog` variables must be global or else they can be collected by lua gc when the script is called again with `RetainLuaState`.

ThistleSifter commented:
> What I like about the underscore is that it should be the outcome of a virtuous cycle: write a loop, get a lint warning, realize that the variable isn't used, and acknowledge it as such.

Not necessarily. It depends on what's more important to the author: knowing what it is, or knowing whether it's intended to be used. Knowing what it is is more important (you can't decide whether it's needed without first knowing what it is), which is why the underscore only makes sense if you *already know* what the value is and you're sure you won't need the reminder (revisting old code a year or 2 later can be an eye opening experience, at which point it may as well have been written by someone else) and you're sure someone else won't need the hint either. Naming things is the best form of documentation we have, and I don't agree with forcing people to name things a certain way just to convince a linter of their intentions. Code is primarily meant to be read and understood by people, especially the person who wrote it, and if the author finds a certain way easier to understand and it isn't causing bugs or opening a potential pitfall, I don't see the issue.



> but if you feel strongly about it I will (along with an ignore hint to make the linter happy).

Thanks, that would be good.

asherber commented:
Okay, @ThistleSifter, I put the `mixin` references back.

As for the other issue, I think the linter rule about flagging unused variables is very useful. In fact, it helped bring to light at least two bugs in the library (one of which was the one you called out, above). Once the linter has flagged a variable like this, there's no rule that says you _have_ to fix it by changing it to an underscore, although I think in most cases it's the clearest and most expedient; you could also choose to leave the variable in place and satisfy the linter by adding an ignore hint to that line of code. 

(I actually agree with just about everything you say in your last comment, especially about readability; it just leads me to a slightly different conclusion!)

rpatters1 commented:
I think my stuff is okay. If it isn't, I'll fix it in a future PR.

rpatters1 commented:
I just realized there is at least one situation where we have to keep unused variables, and that is the `tie.lua` library.

Consider this code:

```lua
local layer, start_note, end_note = tie.calc_tie_span(note, true, true)
```

The `layer` variable may never actually be used in the Lua script, but it is the container for `start_note` and `end_note`. So `layer` is very much under the covers being used by the PDK Framework. It is also keeping the `start_note` and `end_note` instances alive. I'm pretty sure it would be acceptiable to change this to

```lua
local _, start_note, end_note = tie.calc_tie_span(note, true, true)
```

But it is something we need to be aware of.

asherber commented:
@rpatters1 What do you think? This has been open for a week -- time to merge it?

rpatters1 commented:
I've been waiting to merge anything else. lol.

jwink75 commented:
I say we do it… everything I’ve tested seems to work other than the harp wizard, which we commented out and will tackle later. If there are any other problems with my scripts I can always fix later.Sent from my iPhoneOn Jan 30, 2024, at 6:10 PM, rpatters1 ***@***.***> wrote:﻿
I've been waiting to merge anything else. lol.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

rpatters1 commented:
Is there any way to force all the backlog of PRs to run the lint check before we try to merge them?

asherber commented:
Closing and reopening each PR works. I'll take care of it now.
